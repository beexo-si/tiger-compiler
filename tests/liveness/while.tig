/*TIGER_ARGS= -code -flow*/
let
    var i:= 5
in
    while i > 0
    do i := i - 1;
    i
end
