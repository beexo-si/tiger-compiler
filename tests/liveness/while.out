bien!
;;--FRAME--L0:
LABEL: L5
OPER: movq $5, %'d0 D:[T1] S:[]
MOVE: movq %'s0, %'d0 D:T0 S:T1
LABEL: L2
OPER: movq $0, %'d0 D:[T2] S:[]
OPER: cmpq 's0, 's1 D:[] S:[T0,T2]
OPER: jg 'j0 D:[] S:[] J:[L3,L1]
LABEL: L1
MOVE: movq %'s0, %'d0 D:rax S:T0
OPER: jmp 'j0 D:[] S:[] J:[L4]
LABEL: L3
MOVE: movq %'s0, %'d0 D:T3 S:T0
OPER: subq %'d0, $1 D:[T3] S:[T3]
MOVE: movq %'s0, %'d0 D:T0 S:T3
OPER: jmp 'j0 D:[] S:[] J:[L2]
LABEL: L4
OPER:  D:[] S:[rax,rsp,rbp,rbx,r12,r13,r14,r15]
;;-END-FRAME-:
;;--FLOW--L0:
def:
	n0: 
	n1: T1
	n2: T0
	n3: 
	n4: T2
	n5: 
	n6: 
	n7: 
	n8: rax
	n9: 
	n10: 
	n11: T3
	n12: T3
	n13: T0
	n14: 
	n15: 
	n16: 
use:
	n0: 
	n1: 
	n2: T1
	n3: 
	n4: 
	n5: T0, T2
	n6: 
	n7: 
	n8: T0
	n9: 
	n10: 
	n11: T0
	n12: T3
	n13: T3
	n14: 
	n15: 
	n16: rax, rsp, rbp, rbx, r12, r13, r14, r15
ismove:
	n0: false
	n1: false
	n2: true
	n3: false
	n4: false
	n5: false
	n6: false
	n7: false
	n8: true
	n9: false
	n10: false
	n11: true
	n12: false
	n13: true
	n14: false
	n15: false
	n16: false
succesors:
	n0(LABEL: L5): n1
	n1(OPER: movq $5, %'d0 D:[T1] S:[]): n2
	n2(MOVE: movq %'s0, %'d0 D:T0 S:T1): n3
	n3(LABEL: L2): n4
	n4(OPER: movq $0, %'d0 D:[T2] S:[]): n5
	n5(OPER: cmpq 's0, 's1 D:[] S:[T0,T2]): n6
	n6(OPER: jg 'j0 D:[] S:[] J:[L3,L1]): n7, n10
	n7(LABEL: L1): n8
	n8(MOVE: movq %'s0, %'d0 D:rax S:T0): n9
	n9(OPER: jmp 'j0 D:[] S:[] J:[L4]): n15
	n10(LABEL: L3): n11
	n11(MOVE: movq %'s0, %'d0 D:T3 S:T0): n12
	n12(OPER: subq %'d0, $1 D:[T3] S:[T3]): n13
	n13(MOVE: movq %'s0, %'d0 D:T0 S:T3): n14
	n14(OPER: jmp 'j0 D:[] S:[] J:[L2]): n3
	n15(LABEL: L4): n16
	n16(OPER:  D:[] S:[rax,rsp,rbp,rbx,r12,r13,r14,r15]): 
;;-END-FLOW-:
yes!!
